corehq/apps/userreports/reports/specs.py

```diff
@@ -384,42 +383,41 @@ def restricted_to_static(cls, domain):
 
 class SumWhenTemplateColumn(SumWhenColumn):
     type = TypeProperty("sum_when_template")
+    whens = ListProperty(DictProperty)      # List of SumWhenTemplateSpec dicts
 
     @classmethod
     def restricted_to_static(cls, domain):
         return not UCR_SUM_WHEN_TEMPLATES.enabled(domain)
 
     def get_whens(self):
+        from corehq.apps.userreports.reports.factory import SumWhenTemplateFactory
         whens = []
-        for item in self.whens:
-            when, *binds, then = item
-
-            try:
-                expression_class = to_function(when)
-            except ValueError:
-                raise BadSpecError('Badly formatted expression class: {}'.format(when))
-            if not expression_class:
-                raise BadSpecError('Could not find expression class: {}'.format(when))
-            if len(binds) != expression_class.bind_count():
-                raise BadSpecError('Expected {} binds, found {}'.format(len(binds),
-                                                                        expression_class.bind_count()))
-
-            whens.append([expression_class.expression] + binds + [then])
+        for spec in self.whens:
+            template = SumWhenTemplateFactory.make_template(spec)
+            whens.append([template.expression] + template.binds + [template.then])
         return whens
 
 
-class SumWhenTemplate(object):
-    expression = None
+class SumWhenTemplateSpec(JsonObject):
+    type = StringProperty(required=True)
+    expression = StringProperty(required=True)
+    binds = ListProperty()
+    then = IntegerProperty()
```

Comment:
Nice, I like this approach - feels like it'd fail in more obvious ways

Reply:
Yeah, this did turn out better, thanks for the suggestion. I like that end users no longer have to use a list where they might enter [when, then, bind] instead of [when, bind, then].